Add chrome extension docs (#1004)
clerk/clerk-docs: docs/quickstarts/chrome-extension.mdx

---
title: Chrome Extension Quickstart
description: Add authentication and user management to your Chrome Extension with Clerk.
---

# Chrome Extension Quickstart

<TutorialHero
  framework="chrome-extension"
  beforeYouStart={[
    {
      title: "Set up a Clerk application",
      link: "https://clerk.com/docs/quickstarts/setup-clerk",
      icon: "clerk",
    }
  ]}
>
  - Scaffold a new application using the Plasmo framework
  - Install `@clerk/chrome-extension`
  - Set your environment variables
  - Add `<ClerkProvider>` to your application
  - Create a header with Clerk components for users to sign in and out
  - Configure a consistent CRX ID
  - Load your Chrome Extension into your Chromium-based browser
  - Test your Chrome Extension
</TutorialHero>

<Steps>
  ## Configure your authentication options

>   Chrome Extensions can be used just a popup, a side panel, or in conjunction with a web application. Popups and side panels have limited authentication options. [Learn more about what options are available.](/docs/references/chrome-extension/overview#authenication-options)

Reviewer: I hear ya - how's this:

```suggestion
  Chrome Extensions can be used just a popup, a side panel, or in conjunction with a web application. Popups and side panels have limited authentication options. Use the table in [this guide](/docs/references/chrome-extension/overview#authenication-options) to ensure you have the correct authentication options configured in your Clerk Dashboard.
```

  This guide will use a popup.

  ## Scaffold a new application using the Plasmo framework

  [Plasmo](https://docs.plasmo.com/framework) is a feature-rich Browser Extension framework that includes hot reloading, creating development and production extension builds easily from the same code, and more.

  Plasmo strongly recommends using `pnpm`, so this guide will only use `pnpm`-based examples.

  The following command will create an application with Tailwind CSS pre-configured and with a `src/` directory. If desired, you can remove one or both of those options.

  ```bash {{ filename: 'terminal' }}
  pnpm create plasmo --with-tailwindcss --with-src chrome-extension
  cd clerk-chrome-extension
  pnpm dev
  ```

  ## Install `@clerk/chrome-extension`

  Clerk's [Chrome Extension SDK](/docs/references/chrome-extension/overview) gives you access to prebuilt components, React hooks, and helpers to make user authentication easier.

  Add the SDK to your project:

  ```bash {{ filename: 'terminal' }}
  pnpm add @clerk/chrome-extension
  ```

>   ## Set your Clerk API keys

Reviewer: ```suggestion
  ## Set your Clerk publishable key and frontend API URL
```

  Plasmo offers [several options](https://docs.plasmo.com/framework/env) for environment variable files, as the same codebase can be used for development and production builds, as well as for targeting different browsers. This guide will use `.env.development` and `.env.chrome` files.

  <SignedIn>
    Add the following keys to your `.env.development` file. These keys can always be retrieved from the [API Keys](https://dashboard.clerk.com/last-active?path=api-keys) page of your Clerk Dashboard.
  </SignedIn>

  <SignedOut>
    1. Navigate to the [Clerk Dashboard](https://dashboard.clerk.com/last-active?path=api-keys).
    1. In the navigation sidebar, select **API Keys**.

>     1. In the **Quick Copy** section, select **Chrome Extension** and copy your Clerk publishable and frontend API key.

Reviewer: ```suggestion
    1. In the **Quick Copy** section, select **Chrome Extension** and copy your Clerk publishable key and frontend API URL.
```

  ```jsx {{ filename: 'src/popup.tsx', mark: [1, [15, 24], 26] }}
  import {
    ClerkProvider,
    SignInButton,
    SignedIn,
    SignedOut,
    UserButton,
  } from '@clerk/chrome-extension'
  import { CountButton } from '~features/count-button'

  import '~style.css'

  const PUBLISHABLE_KEY = process.env.PLASMO_PUBLIC_CLERK_PUBLISHABLE_KEY

  if (!PUBLISHABLE_KEY) {
    throw new Error('Please add the PLASMO_PUBLIC_CLERK_PUBLISHABLE_KEY to the .env.development file')
  }

  function IndexPopup() {
    return (
      <ClerkProvider publishableKey={PUBLISHABLE_KEY}>
        <div className="plasmo-flex plasmo-items-center plasmo-justify-center plasmo-h-[600px] plasmo-w-[800px] plasmo-flex plasmo-flex-col">
          <header className="plasmo-w-full">
            <SignedOut>
              <SignInButton mode="modal" />
            </SignedOut>
            <SignedIn>
              <UserButton />
            </SignedIn>
          </header>
          <main className="plasmo-grow">
            <CountButton />
          </main>
        </div>
      </ClerkProvider>
    )
  }

  export default IndexPopup
  ```

  ## Create a consistent CRX ID for your extension

>   Chrome Extensions have a unique CRX ID that rotates by default, which can cause errors and issues with the Clerk integration. To avoid these problems, ensure that you have a **consistent** CRX ID in both development and production for your extension by following these steps:

Reviewer: Is this correct? It rotates by default.

Author: Yes -- that's why we create the consistent one. Otherwise you end having it change, not realizing it and Clerk stops working.

Reviewer: okay awesome - just wanted to make sure I wrote accurate info!

  In your `package.json`, add `"key": "$CRX_PUBLIC_KEY"` to the `manifest` object, as shown in the following example:

  ```bash {{ filename: 'package.json', mark: [33] }}
  {
    "name": "chrome-extension",
    "displayName": "Chrome extension",
    "version": "0.0.1",
    "description": "A basic implementation of Clerk with a Plasmo-based Chrome Extension.",
    "author": "Clerk",
    "scripts": {
      "dev": "plasmo dev",
      "build": "plasmo build",
      "package": "plasmo package"
    },
    "dependencies": {
      "@clerk/chrome-extension": "1.1.5-snapshot.va3e1056",
      "plasmo": "0.88.0",
      "react": "18.2.0",
      "react-dom": "18.2.0",
      "tailwindcss": "3.4.1"
    },
    "devDependencies": {
      "@ianvs/prettier-plugin-sort-imports": "4.1.1",
      "@types/chrome": "0.0.258",
      "@types/node": "20.11.5",
      "@types/react": "18.2.48",
      "@types/react-dom": "18.2.18",
      "postcss": "8.4.33",
      "prettier": "3.2.4",
      "typescript": "5.3.3"
    },
    "manifest": {
      "host_permissions": [
        "https://*/*"
      ],
      "key": "$CRX_PUBLIC_KEY"
    }
  }
  ```

  ## Load your Chrome Extension into your Chromium-based browser

>   The CRXJS plugin that you installed facilitates Chrome Extension development by automatically "hot loading" the application while it runs. This ensures the `build/chrome-mv3-dev` folder remains up to date. Without the plugin, you would need to manually execute the build command and reload your Chrome Extension after each change. Plasmo automates this process, streamlining development.

Reviewer: The guide never mentions installing this plugin - I thought Plasmo was going to be used instead?

Author: Yup! I missed this -- will update.

Reviewer: okay once that's all correctly updated I can finish testing and reviewing!

---

# Chrome Extension Quickstart

<TutorialHero
  framework="nextjs"
  beforeYouStart={[
    {
      title: "Set up a Clerk application",
      link: "https://clerk.com/docs/quickstarts/setup-clerk",
      icon: "clerk",
    },
    {
      title: "Create a React application with Vite",
      link: "https://vitejs.dev/guide/#scaffolding-your-first-vite-project",
      icon: "nextjs",
    },
    {
      title: "Install the CRXJS Plugin",
      link: "https://crxjs.dev/vite-plugin/getting-started/react/create-project",
      icon: "nextjs",
    }
  ]}
>

- Install `@clerk/chrome-extension`
- Set your environment variables
- Add `<ClerkProvider>` to your application
- Create a header with Clerk components for users to sign in and out
- Add basic styles

</TutorialHero>

This tutorial only supports the email & password, email & email code, and SMS & SMS code [authentication options](/docs/authentication/configuration/sign-up-sign-in-options). Email & email link, social sign in (OAuth), SAML, and invites are not supported.

<Steps>

### Install `@clerk/chrome-extension`

{/* TODO: Fill the link the reference */}

> Clerk's [Chrome extension SDK](#) gives you access to prebuilt components, React hooks, and helpers to make user authentication easier.

Reviewer: ```suggestion
Clerk's [Chrome Extension SDK](#) gives you access to prebuilt components, React hooks, and helpers to make user authentication easier.
```

> Clerk's [Chrome Extension SDK](#) gives you access to prebuilt components, React hooks, and helpers to make user authentication easier.

Author: Is this a placeholder? Or is it meant to link to itself?

Reviewer: yes there is a TODO right above it that says to fill the link once we create the reference in our docs.

Author: Yes, I totally missed that.

Author: Added link.

Add the SDK to your project:

<CodeBlockTabs type="installer" options={["npm", "yarn",  "pnpm" ]}>
  ```bash filename="terminal"
  npm install @clerk/chrome-extension
  ```
  ```bash filename="terminal"
  yarn add @clerk/chrome-extension
  ```

  ```bash filename="terminal"
  pnpm add @clerk/chrome-extension
  ```
</CodeBlockTabs>

### Set your environment variables

<SignedIn>

> Add the following keys to your `.env.local` file. These keys can always be retrieved from the [API Keys](https://dashboard.clerk.com/last-active?path=api-keys) page of your Clerk Dashboard.

Author: ```suggestion
Add the following keys to your `.env` file. These keys can always be retrieved from the [API Keys](https://dashboard.clerk.com/last-active?path=api-keys) page of your Clerk Dashboard.
```

`.env.local` is supported in Vite but `.env` seems to be the preferred method by the community.

- [`<SignedIn>`](/docs/components/control/signed-in): Children of this component can only be seen while **signed in**.
- [`<SignedOut>`](/docs/components/control/signed-out): Children of this component can only be seen while **signed out**.
- [`<UserButton />`](/docs/components/user/user-button): A prebuilt component that comes styled out of the box to show the avatar from the account the user is signed in with.
- [`<SignInButton />`](/docs/components/unstyled/sign-in-button): An unstyled component that links to the sign-in page or displays the sign-in modal.

```jsx filename="src/App.tsx" {2-7, 12-19}
import "./App.css";
import {
  SignInButton,
  SignedIn,
  SignedOut,
  UserButton,
} from "@clerk/chrome-extension";

function App() {
  return (
    <>
      <header className="header">
        <SignedOut>
          <SignInButton mode="modal" />
        </SignedOut>
        <SignedIn>
          <UserButton />
        </SignedIn>
      </header>
    </>
  );
}

export default App;
```

### Add basic styles (optional)

> To make your development experience easier, the following example includes some basic styles for your application. You can customize these styles to fit your application's design.

Reviewer: ```suggestion
To make your development experience easier, the following examples include some basic styles for your application. You can customize these styles to fit your application's design.
```

    }
  ]}
>
  - Scaffold a new application using the Plasmo framework
  - Install `@clerk/chrome-extension`
  - Set your environment variables
  - Add `<ClerkProvider>` to your application
  - Create a header with Clerk components for users to sign in and out
  - Configure a consistent CRX ID
  - Load your Chrome Extension into your Chromium-based browser
  - Test your Chrome Extension
</TutorialHero>

<Steps>
  ## Configure your authentication options

>   Chrome Extensions can be used just a popup, a side panel, or in conjunction with a web application. Popups and side panels have limited authentication options. [Learn more about what options are available.](/docs/references/chrome-extension/overview#authenication-options)

Author: I'm not sure I like removing the following, as I think this will be a confusing area for customers and I want to really draw attention to it. Happy to talk about options too -- I just want to make sure this is highlighted enough. Do you feel the revised copy is really strong on this point?

---
Use the table to ensure you have the correct authentication options enabled/disabled in your [Clerk Dashboard](https://dashboard.clerk.com/last-active?path=authentication).

pnpm add @clerk/chrome-extension
```

> ## Set your Clerk API keys

Author: This should probably keep the Frontend API URL as the user is copying one key and then one URL, versus keys like with most frameworks.

      </ClerkProvider>
    )
  }

  export default IndexPopup
  ```

  ### Create a header with Clerk components

  You can control what content signed in and signed out users can see with Clerk's [prebuilt components](/docs/components/overview). Create a header with the following Clerk components. (With Chrome Extensions, you can also add this logic to a footer).

  - [`<SignedIn>`](/docs/components/control/signed-in): Children of this component can only be seen while **signed in**.
  - [`<SignedOut>`](/docs/components/control/signed-out): Children of this component can only be seen while **signed out**.
  - [`<UserButton />`](/docs/components/user/user-button): A prebuilt component that comes styled out-of-the-box to show the avatar from the account the user is signed in with.

>   - [`<SignInButton />`](/docs/components/unstyled/sign-in-button): An unstyled component that links to the sign-in page. For this example, because you have not specified any props or [environment variables](/docs/deployments/clerk-environment-variables) for the sign-in URL, the component will link to the [Account Portal sign-in page.](/docs/customization/account-portal/overview)

Author: The sample code is using a modal. The changes here should be reverted.

Reviewer: ```suggestion
  - [`<SignInButton />`](/docs/components/unstyled/sign-in-button): An unstyled component that links to the sign-in page or opens a modal. For this example, because you have not specified any props or [environment variables](/docs/deployments/clerk-environment-variables) for the sign-in URL, the component will link to the [Account Portal sign-in page.](/docs/customization/account-portal/overview)
```

Author: I would still drop the 'sign-in page' as the Quickstart doesn't set that up.

    )
  }

  export default IndexPopup
  ```

  ### Create a consistent CRX ID for your extension

  Chrome Extensions have a unique CRX ID that rotates by default, which can cause errors with the Clerk integration. To avoid these problems, ensure that you have a **consistent** CRX ID in both development and production for your extension by following these steps:

  1. Visit Plasmo Itero's [Generate Keypairs](https://itero.plasmo.com/tools/generate-keypairs) tool.
  1. Select **Generate KeyPairs**.
  1. Save the **Private Key** somewhere secure in case you need it in the future. Save the **Public Key** and the **CRX ID** for the next steps.

  ### Create an `.env.chrome` file to store your public key

  Create an `.env.chrome` file and add your public key to it, as shown in the following example:

  ```env {{ filename: '.env.chrome' }}
  CRX_PUBLIC_KEY=<YOUR_PUBLIC_KEY>
  ```

  ### Edit your `package.json` to use the new public key

>   Plasmo uses the `package.json` to generate a `manifest.json` on build, and allows for the use of environment variables in `package.json`.

Author: Thoughts on linking to https://docs.plasmo.com/framework#where-is-the-manifestjson-file here, as the info about this was removed?

Reviewer: ```suggestion
  Plasmo [uses the `package.json` to generate a `manifest.json` on build](https://docs.plasmo.com/framework#where-is-the-manifestjson-file), and allows for the use of environment variables in `package.json`.
```
